src/GafferImage/CreateViews.cpp

```diff
@@ -114,59 +113,145 @@ const Gaffer::ArrayPlug *CreateViews::viewsPlug() const
 }
 
-const Gaffer::NameSwitch *CreateViews::nameSwitchNode() const
+Gaffer::Switch *CreateViews::switchNode()
 {
-	return getChild<Gaffer::NameSwitch>( g_firstPlugIndex + 1 );
+	return getChild<Gaffer::Switch>( g_firstPlugIndex + 2 );
 }
 
+const Gaffer::Switch *CreateViews::switchNode() const
+{
+	return getChild<Gaffer::Switch>( g_firstPlugIndex + 2 );
+}
 
 void CreateViews::synchronizeSwitch()
 {
 	int numViews = viewsPlug()->children().size();
-	nameSwitchNode()->inPlugs()->resize( 1 + numViews );
+	switchNode()->inPlugs()->resize( numViews );
 	for( int i = 0; i < numViews; i++ )
 	{
 		NameValuePlug *source = viewsPlug()->getChild<NameValuePlug>( i );
-		NameValuePlug *dest = nameSwitchNode()->inPlugs()->getChild<NameValuePlug>( i + 1 );
-		dest->namePlug()->setInput( source->namePlug() );
-		dest->enabledPlug()->setInput( source->enabledPlug() );
-		dest->valuePlug()->setInput( source->valuePlug() );
+		ImagePlug *dest = switchNode()->inPlugs()->getChild<ImagePlug>( i );
+		dest->setInput( source->valuePlug() );
 	}
 }
 
 void CreateViews::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
 {
 	ImageNode::affects( input, outputs );
 
-	auto nameValuePlug = input->parent<NameValuePlug>();
+	auto nameValuePlug = input->ancestor<NameValuePlug>();
 	if(
 		nameValuePlug && nameValuePlug->parent() == viewsPlug() &&
-		( input == nameValuePlug->namePlug() || input == nameValuePlug->enabledPlug() )
+		( input == nameValuePlug->namePlug() || input == nameValuePlug->enabledPlug() || input == nameValuePlug->valuePlug<ImagePlug>()->viewNamesPlug() )
 	)
 	{
 		outputs.push_back( outPlug()->viewNamesPlug() );
+		outputs.push_back( indexPlug() );
 	}
 
 }
 
-void CreateViews::hashViewNames( const GafferImage::ImagePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+void CreateViews::hash( const Gaffer::ValuePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
 {
-	ImageNode::hashViewNames( output, context, h );
-	for( const GraphComponentPtr &p : viewsPlug()->children() )
+	if( output != indexPlug() )
+	{
+		ImageNode::hash( output, context, h );
+		return;
+	}
+
+
+	h.append( context->variableHash( ImagePlug::viewNameContextName ) );
+
+	Context::EditableScope s( context );
+	s.remove( ImagePlug::viewNameContextName );
+
+	for( auto &nameValue : NameValuePlug::Range( *viewsPlug() ) )
+	{
+		if( nameValue->enabledPlug() )
+		{
+			nameValue->enabledPlug()->hash( h );
+		}
+		nameValue->namePlug()->hash( h );
+		nameValue->valuePlug<ImagePlug>()->viewNamesPlug()->hash( h );
+	}
+}
+
+void CreateViews::compute( Gaffer::ValuePlug *output, const Gaffer::Context *context ) const
+{
+	if( output != indexPlug() )
+	{
+		ImageNode::compute( output, context );
+		return;
+	}
+
+	const std::string &currentView = context->get<std::string>( ImagePlug::viewNameContextName );
+
+	Context::EditableScope s( context );
+	s.remove( ImagePlug::viewNameContextName );
```

Comment:
I think the call to `remove()` is invalidating the reference we take on L197.

Reply:
Daniel pointed out that this is fine, because the removal is from a new EditableScope, not the original context.